спринт №16. Все экрана плеера переделаны в архитектурном стиле MVVM

amdoit-git opened the pull request.

amdoit-git committed:
Экран плеера перенесен в пакет Player с корня приложения.
Добавлены пакеты под другие экраны

amdoit-git committed:
Экран плеера переделан в стиле архитектуры MVVM

amdoit-git committed:
Решена проблема срабатывания setOnCheckedChangeListener без действий со стороны пользователя

amdoit-git committed:
Добавлен класс обертка LiveDataWithStartDataSet над MutableLiveData, который позволяет передавать разные типы данных через одну MutableLiveData
Переделан экран плеера с использованием MutableLiveData

amdoit-git committed:
Добавлен синхронный поиск через интерактор запускаемый в отдельном потоке

amdoit-git committed:
Экран поска по трекам сделан в стиле MVVM. Но с поехавшей версткой из-за перехода на View Binding

amdoit-git committed:
Исправлена поехавшая верстка из-за перехода на View Binding
Добавлено перемещение трека в ТОПе на первое место при его выборе в истории поиска

amdoit-git committed:
Общие данные и интерфейсы перенесены в пакет common.
Удален больше не используемый класс MusicPlayer.kt

amdoit-git committed:
Экран настроек приложения переделан в архитектурном стиле MVVM

amdoit-git committed:
Медиаплеер после установки нового url ресурса не сбрасывал состояние ошибки подготовки трека. Ошибка устранена

amdoit-git committed:
Убрана передача контекста и Shared Preferences через App

amdoit-git committed:
Удалены неиспользуемые классы оставшиеся в проекте со времен до чистой архитектуры

amdoit-git committed:
Интерфес поиска на iTunes вынесен из пакета dto

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/data/SearchHistory.kt:
```diff
@@ -11,13 +11,13 @@ class SearchHistory {
 
     companion object {
```
Лучше весь SearchHistory сделать как object. Тогда можно убрать companion object

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/data/impl/AppSettingsRepositoryImpl.kt:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    override fun saveTheme(theme: AppTheme) {
+        sharedPrefs.edit().putBoolean(DARK_THEME_KEY, theme.dark).apply()
```
![SharedPreferences](https://github.com/user-attachments/assets/4e5e91cd-24e8-40d5-bf5e-1834a0b0ffd8)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/data/impl/TracksHistoryRepositoryImpl.kt:
```diff
@@ -1,10 +1,19 @@
+import com.example.playlistmaker.common.domain.repository.TracksHistoryRepository
 
-class TracksHistoryRepositoryImpl() : TracksHistoryRepository {
+class TracksHistoryRepositoryImpl(context: Context) : TracksHistoryRepository {
```
Лучше сразу передать готовый SharedPreferences как зависимость.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/domain/consumer/Consumer.kt:
```diff
@@ -0,0 +1,5 @@
+package com.example.playlistmaker.common.domain.consumer
+
+interface Consumer<T> {
```
Используй SAM interface
https://kotlinlang.org/docs/fun-interfaces.html

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/presentation/LiveDataWithStartDataSet.kt:
```diff
@@ -0,0 +1,53 @@
+import androidx.lifecycle.Observer
+import java.util.concurrent.atomic.AtomicBoolean
+
+class LiveDataWithStartDataSet<T : Any> : MutableLiveData<T>() {
```
Рекомендую посмотреть в строну корутин, там уже решена эта проблема.
Например так:
private val flow = MutableSharedFlow<T>(replay = 0, extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/presentation/SCREEN_NAME.kt:
```diff
@@ -0,0 +1,14 @@
+import com.example.playlistmaker.search.ui.SearchActivity
+import com.example.playlistmaker.settings.ui.SettingsActivity
+
+enum class SCREEN_NAME(val className: Class<out Activity>) {
```
1) SCREEN_NAME переименуй в ScreenName. Имена классов в Kotlin пишутся в CamelCase нотации.
2) Не рекомендую изобретать собственную навигацию. Через пару спринтов, все это надо будет удалить, т.к. будет другая навигация.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/common/presentation/SingleEventLiveData.kt:
```diff
@@ -0,0 +1,28 @@
+import androidx.lifecycle.Observer
+import java.util.concurrent.atomic.AtomicBoolean
+
+class SingleEventLiveData<T> : MutableLiveData<T>() {
```
Хорошо, что залез под капот MutableLiveData, но не рекомендую изобретать свои решения, т.к. Эта проблема уже решена в корутинах.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/creator/Creator.kt:
```diff
@@ -1,29 +1,66 @@
     }
 
-    fun provideMediaPlayerRepository(): MediaPlayerRepository{
+    fun provideMediaPlayerRepository(): MediaPlayerRepository {
```
Сделать приватным

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/creator/Creator.kt:
```diff
@@ -1,29 +1,66 @@
 
-    fun provideTracksHistoryRepository(): TracksHistoryRepository{
-        return TracksHistoryRepositoryImpl()
+    fun provideTracksHistoryRepository(context: Context): TracksHistoryRepository {
```
Сделать приватным

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/creator/Creator.kt:
```diff
@@ -1,29 +1,66 @@
+        return ITunesInteractorImpl(provideITunesRepository(context))
+    }
+
+    fun provideITunesRepository(context: Context): ITunesRepository {
```
Сделать приватным

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/creator/Creator.kt:
```diff
@@ -1,29 +1,66 @@
+        return ExternalNavigatorInteractorImpl(provideExternalNavigatorRepository(context))
+    }
+
+    fun provideExternalNavigatorRepository(context: Context): ExternalNavigatorRepository {
```
Сделать приватным

amdoit-git committed:
Добавлена отмена поискового запроса при очистке посковой строки и показе истории поиска

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/main/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,27 @@
+
+        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+
+        application.applicationContext.startActivity(intent)
```
Запускать активити эта задача самой активити, а не ViewModel. Рекомендую перенести код в активити.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/main/ui/MainActivity.kt:
```diff
@@ -0,0 +1,38 @@
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var presenter: MainViewModel
```
ViewModel это не presenter, лучше переменную переименовать в viewModel

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/main/ui/MainActivity.kt:
```diff
@@ -0,0 +1,38 @@
+            ViewModelProvider(this, MainViewModel.Factory(application))[MainViewModel::class.java]
+
+        binding.buttonSearch.setOnClickListener {
+            presenter.openScreen(SCREEN_NAME.SEARCH)
```
Не надо переносить старт активити во viewModel.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/player/data/MediaPlayerService.kt:
```diff
@@ -1,10 +1,10 @@
-import com.example.playlistmaker.domain.models.Track
+import android.os.SystemClock
 
 class MediaPlayerService {
```
Не следует писать весь код класса в companion object.
Если тебе нужен singleton, то сделай object MediaPlayerService, вместо класса.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/player/presentation/PlayerScreenViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class PlayerScreenViewModel(private val application: Application, jsonTrack: String) :
```
Публичные и приватные методы в перемешку.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/player/ui/PlayerScreenActivity.kt:
```diff
@@ -0,0 +1,165 @@
+
+                        it.duration?.let {
+
+                        }
```
Пустой кусок кода.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/player/ui/PlayerScreenActivity.kt:
```diff
@@ -0,0 +1,165 @@
+                        binding.infoText.text = it.message
+
+                        binding.info.visibility = if (it.isVisible) {
+                            View.VISIBLE
```
![view](https://github.com/user-attachments/assets/db7a2535-7873-4f69-a745-c1465c045db6)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/player/ui/PlayerScreenActivity.kt:
```diff
@@ -0,0 +1,165 @@
+
+                    is PlayerScreenData.TrackData -> fillTrackInfo(it.track)
+
+                    is PlayerScreenData.TrackProgress -> {
```
Сделай начальное значение для TrackProgress, чтобы потом не пришлось "бороться" с null

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/search/ui/SearchActivity.kt:
```diff
@@ -0,0 +1,208 @@
+
+    private fun initTextField(editText: EditText) {
+
+        editText.addTextChangedListener(object : TextWatcher {
```
![TextWatcher](https://github.com/user-attachments/assets/80a42da6-15d6-407d-af1b-756933aa9c29)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/search/ui/TrackAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    companion object {
```
Весь companion object можно сделать private

amdoit-git committed:
MediaPlayerService и SearchHistory из классов с companion object переделаны в глобальные объекты
Публичные методы там где не требуется обращение к ним из вне заменены на приватные
presenter переименован в viewModel
